File: CON BDD/EUREKABANK_SOAP_DOTNET_GR06/05. CLIESCRITORIO/Controlador/BancoController.cs
using CLIESCRITORIO.Servicio;

namespace CLIESCRITORIO.Controlador
{
    public class BancoController
    {
        private readonly SoapClient _soapClient;
        private readonly Sesion _sesion = new();

        public BancoController(SoapClient? soapClient = null)
        {
            _soapClient = soapClient ?? new SoapClient();
        }

        public Sesion Sesion => _sesion;

        public bool Login(string usuario, string clave)
        {
            if (!_soapClient.IniciarSesion(usuario, clave))
            {
                return false;
            }

            _sesion.Usuario = usuario;
            _sesion.Admin = string.Equals(usuario, "monster", StringComparison.OrdinalIgnoreCase);
            _sesion.ClienteAsignado = _sesion.Admin ? string.Empty : _soapClient.ClienteDeUsuario(usuario);
            _sesion.Cuentas = new List<CuentaResumen>();

            if (!_sesion.Admin)
            {
                CargarCuentas(_sesion.ClienteAsignado);
            }

            return true;
        }

        public void Logout()
        {
            _sesion.Usuario = null;
            _sesion.Admin = false;
            _sesion.ClienteAsignado = string.Empty;
            _sesion.Cuentas = new List<CuentaResumen>();
        }

        public List<ClienteResumen> ListarClientes() => _soapClient.ListarClientes();

        public void CargarCuentas(string criterio)
        {
            var c = _sesion.Admin ? criterio : _sesion.ClienteAsignado;
            _sesion.Cuentas = _soapClient.ListarCuentasPorCliente(c) ?? new List<CuentaResumen>();
        }

        public List<CuentaResumen> GetCuentas() => _sesion.Cuentas;

        public double SaldoTotal() => _sesion.Cuentas.Sum(c => c.Saldo);

        private static Resultado Denegado(string msg) => new()
        {
            Exitoso = false,
            Mensaje = msg
        };

        public Resultado ConsultarSaldo(string cuenta)
        {
            if (!_sesion.CuentaPropia(cuenta))
            {
                return Denegado($"No tienes acceso a la cuenta {cuenta}.");
            }

            return _soapClient.ConsultarSaldo(cuenta);
        }

        public Resultado Depositar(string cuenta, string monto, string moneda)
        {
            if (!_sesion.Admin)
            {
                return Denegado("Solo el administrador puede depositar. Usa transferencia.");
            }

            if (!_sesion.CuentaPropia(cuenta))
            {
                return Denegado($"No tienes acceso a la cuenta {cuenta}.");
            }

            return _soapClient.Depositar(cuenta, monto, moneda);
        }

        public Resultado Retirar(string cuenta, string monto, string moneda)
        {
            if (!_sesion.CuentaPropia(cuenta))
            {
                return Denegado($"No tienes acceso a la cuenta {cuenta}.");
            }

            return _soapClient.Retirar(cuenta, monto, moneda);
        }

        public Resultado Transferir(string origen, string destino, string monto, string moneda)
        {
            if (!_sesion.CuentaPropia(origen))
            {
                return Denegado($"No tienes acceso a la cuenta de origen {origen}.");
            }

            return _soapClient.Transferir(origen, destino, monto, moneda);
        }

        public List<MovimientoModel> Movimientos(string cuenta)
        {
            if (!_sesion.CuentaPropia(cuenta))
            {
                return new List<MovimientoModel>();
            }

            return _soapClient.ListarMovimientos(cuenta);
        }

        public Resultado RegistrarCliente(string paterno, string materno, string nombre,
            string dni, string ciudad, string direccion, string telefono, string email)
        {
            if (!_sesion.Admin)
            {
                return Denegado("Solo el administrador puede registrar clientes.");
            }

            return _soapClient.RegistrarCliente(paterno, materno, nombre, dni, ciudad, direccion, telefono, email);
        }

        public Resultado RegistrarCuenta(string cliente, string moneda)
        {
            if (!_sesion.Admin)
            {
                return Denegado("Solo el administrador puede registrar cuentas.");
            }

            return _soapClient.RegistrarCuenta(cliente, moneda);
        }

        public Resultado EliminarCuenta(string cuenta)
        {
            if (!_sesion.Admin)
            {
                return Denegado("Solo el administrador puede eliminar cuentas.");
            }

            return _soapClient.EliminarCuenta(cuenta);
        }
    }
}

File: CON BDD/EUREKABANK_RESTFULL_JAVA_GR06/05. CLIESCRITORIO/eurekabank_restful_java_cliesc_gr06/src/main/java/ec/edu/monster/controlador/BancoController.java
package ec.edu.monster.controlador;

import ec.edu.monster.servicio.CuentaClient;
import ec.edu.monster.servicio.LoginClient;
import ec.edu.monster.servicio.MovimientoClient;
import ec.edu.monster.servicio.Sesion;
import ec.edu.monster.ws.ClienteResumen;
import ec.edu.monster.ws.CuentaResumen;
import ec.edu.monster.ws.MovimientoModel;
import ec.edu.monster.ws.Resultado;
import java.util.List;

/**
 * Controlador (MVC) de la consola: orquesta servicios y aplica las MISMAS
 * reglas que el cliente web (rol admin/cliente, guard de cuenta propia,
 * depósito solo admin, conversión de moneda).
 */
public class BancoController {

    private final LoginClient loginClient = new LoginClient();
    private final CuentaClient cuentaClient = new CuentaClient();
    private final MovimientoClient movimientoClient = new MovimientoClient();
    private final Sesion sesion = new Sesion();

    public Sesion getSesion() { return sesion; }

    /* ---------- Sesión ---------- */

    public boolean login(String usuario, String clave) {
        if (!loginClient.iniciarSesion(usuario, clave)) {
            return false;
        }
        sesion.setUsuario(usuario);
        boolean admin = "monster".equalsIgnoreCase(usuario);
        sesion.setAdmin(admin);
        sesion.setClienteAsignado(admin ? "" : loginClient.clienteDeUsuario(usuario));
        sesion.setCuentas(null);
        if (!admin) {
            cargarCuentas(sesion.getClienteAsignado());
        }
        return true;
    }

    public void logout() {
        sesion.setUsuario(null);
        sesion.setAdmin(false);
        sesion.setClienteAsignado("");
        sesion.setCuentas(null);
    }

    /* ---------- Clientes / cuentas ---------- */

    public List<ClienteResumen> listarClientes() {
        return cuentaClient.listarClientes();
    }

    /** Carga las cuentas. Si no es admin, se fuerza su cliente asignado. */
    public void cargarCuentas(String criterio) {
        String c = sesion.isAdmin() ? criterio : sesion.getClienteAsignado();
        sesion.setCuentas(cuentaClient.listarCuentasPorCliente(c));
    }

    public List<CuentaResumen> getCuentas() { return sesion.getCuentas(); }

    public double saldoTotal() {
        double t = 0;
        for (CuentaResumen c : sesion.getCuentas()) t += c.getSaldo();
        return t;
    }

    /* ---------- Operaciones ---------- */

    private Resultado denegado(String msg) {
        Resultado r = new Resultado();
        r.setExito(false);
        r.setMensaje(msg);
        return r;
    }

    public Resultado consultarSaldo(String cuenta) {
        if (!sesion.cuentaPropia(cuenta)) {
            return denegado("No tienes acceso a la cuenta " + cuenta + ".");
        }
        return cuentaClient.consultarSaldo(cuenta);
    }

    public Resultado depositar(String cuenta, String monto, String moneda) {
        if (!sesion.isAdmin()) {
            return denegado("Solo el administrador puede depositar. Usa transferencia.");
        }
        if (!sesion.cuentaPropia(cuenta)) {
            return denegado("No tienes acceso a la cuenta " + cuenta + ".");
        }
        return cuentaClient.depositar(cuenta, monto, moneda);
    }

    public Resultado retirar(String cuenta, String monto, String moneda) {
        if (!sesion.cuentaPropia(cuenta)) {
            return denegado("No tienes acceso a la cuenta " + cuenta + ".");
        }
        return cuentaClient.retirar(cuenta, monto, moneda);
    }

    public Resultado transferir(String origen, String destino, String monto, String moneda) {
        if (!sesion.cuentaPropia(origen)) {
            return denegado("No tienes acceso a la cuenta de origen " + origen + ".");
        }
        return cuentaClient.transferir(origen, destino, monto, moneda);
    }

    public List<MovimientoModel> movimientos(String cuenta) {
        if (!sesion.cuentaPropia(cuenta)) {
            return java.util.Collections.emptyList();
        }
        return movimientoClient.listarMovimientos(cuenta);
    }

    /* ---------- Admin ---------- */

    public Resultado registrarCliente(String paterno, String materno, String nombre,
            String dni, String ciudad, String direccion, String telefono, String email) {
        if (!sesion.isAdmin()) {
            return denegado("Solo el administrador puede registrar clientes.");
        }
        return cuentaClient.registrarCliente(paterno, materno, nombre, dni,
                ciudad, direccion, telefono, email);
    }

    public Resultado registrarCuenta(String cliente, String moneda) {
        if (!sesion.isAdmin()) {
            return denegado("Solo el administrador puede registrar cuentas.");
        }
        return cuentaClient.registrarCuenta(cliente, moneda);
    }

    public Resultado eliminarCuenta(String cuenta) {
        if (!sesion.isAdmin()) {
            return denegado("Solo el administrador puede eliminar cuentas.");
        }
        return cuentaClient.eliminarCuenta(cuenta);
    }
}
